src/main/scala/lms/core/stub.scala

```diff
@@ -80,6 +80,15 @@ object Adapter extends FrontEnd {
 
   class MyGraphBuilder extends GraphBuilder {
 
+    override def gatherEffectDepsWrite(s: String, as: Seq[Def], lw: Sym, lr: Seq[Sym]): (Seq[Sym], Seq[Sym]) =
+    findDefinition(latest(lw)) match {
+      case Some(Node(_, "array_set", as2, deps)) if (s == "array_set" && as.init == as2.init) =>
+        // If latest(lw) is setting the same array at the same index, we do not add hard dependence but soft dependence
+        // In addition, we need to inherite the soft and hard deps of latest(lw)
+        (deps.sdeps.toSeq ++ Seq(latest(lw)), deps.hdeps.toSeq)
```

Comment:
the order of dependencies doesn't matter: latest(lw) +: deps.sdeps.toSeq.
Why not keeping a Set?

Reply:
make sense. will fix :)

Reply:
fixed :)